ons_alpha/private_media/models.py

```diff
@@ -0,0 +1,217 @@
+import logging
+
+from collections.abc import Iterable
+
+from botocore.exceptions import ClientError
+from django.core.files.base import File
+from django.db import models
+from django.utils import timezone
+from wagtail.models import Collection, CollectionMember
+
+from ons_alpha.private_media.utils import (
+    get_private_root_collection_id,
+    get_private_root_collection_path,
+    get_public_root_collection_id,
+    get_public_root_collection_path,
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+class ProtectedCollection(models.Model):
+    collection = models.OneToOneField(Collection, on_delete=models.PROTECT)
+
+    def __str__(self):
+        return f"Exists to prevent accidental deletion of {self.collection!r}"
+
+
+class PrivateMediaMixin(models.Model):
+    is_private = models.BooleanField(default=False, editable=False)
+    last_public_collection = models.ForeignKey(
+        Collection, null=True, editable=False, on_delete=models.SET_NULL, related_name="+"
+    )
+    last_private_collection = models.ForeignKey(
+        Collection, null=True, editable=False, on_delete=models.SET_NULL, related_name="+"
+    )
+    privacy_last_changed = models.DateTimeField(null=True, editable=False)
+    acls_last_set = models.DateTimeField(null=True, editable=False)
+
+    privacy_fields = (
+        "is_private",
+        "last_public_collection",
+        "last_private_collection",
+        "privacy_last_changed",
+        "acls_last_changed",
+    )
+
+    class Meta:
+        abstract = True
+
+    def save(self, *args, set_file_acls: bool = True, **kwargs) -> None:
+        privacy_changed = self.determine_privacy()
+        # Save model field changes at this point
+        super().save(*args, **kwargs)
+        # Trigger ACL updates after-the-fact
+        if privacy_changed and set_file_acls:
+            success = self.set_file_acls("private" if self.is_private else "public-read")
+            if success:
+                self.save(update_fields=["acls_last_set"], set_file_acls=False)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.was_private_on_init = self.is_private
+
+    @property
+    def is_public(self) -> bool:
+        return not self.is_private
+
+    def determine_privacy(self) -> bool:
+        privacy_changed = False
+        self.is_private = self.collection.path.startswith(get_private_root_collection_path())
+        if self.is_private and not self.was_private_on_init:
```

Comment:
Nit: `was_private_on_init` seems to only be used here. It might be easier to store `was_private_on_init` in the function before setting `is_private`, since it's only read here to compare to the previous value of `is_private`